[HLSL] Implement support for HLSL intrinsic  - saturate

bharadwajy opened the pull request:
Implement support for HLSL intrinsic saturate.
Implement DXIL codegen for the intrinsic saturate by lowering it to DXIL Op dx.saturate.
Implement SPIRV codegen by transforming saturate(x) to clamp(x, 0.0f, 1.0f).

Add tests for DXIL and SPIRV CodeGen.

Commit:
Implement support to compile HLSL intrinsic "saturate" to DXIL

Add SPIRV Codegen support to transform saturate(x) to clamp(x, 0.0, 1.0)
Add tests for DXIL and SPIRV CodeGen.

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18666,6 +18666,15 @@ case Builtin::BI__builtin_hlsl_elementwise_isinf: {
   }
+  case Builtin::BI__builtin_hlsl_elementwise_saturate: {
+    Value *Op0 = EmitScalarExpr(E->getArg(0));
+    if (!E->getArg(0)->getType()->hasFloatingRepresentation())
```
you already handle this case in `clang/lib/Sema/SemaHLSL.cpp` this should just be an assert.

farzonl commented on clang/test/CodeGenHLSL/builtins/saturate.hlsl:
```diff
@@ -0,0 +1,54 @@
+// RUN:   --check-prefixes=CHECK,NATIVE_HALF
+// RUN: %clang_cc1 -std=hlsl2021 -finclude-default-header -x hlsl -triple \
+// RUN:   dxil-pc-shadermodel6.3-library %s -emit-llvm -disable-llvm-passes \
+// RUN:   -o - | FileCheck %s --check-prefixes=CHECK,NO_HALF
```
add spirv tests so you can check for `@llvm.spv.saturate`

farzonl commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -362,6 +364,34 @@ static bool expandClampIntrinsic(CallInst *Orig, Intrinsic::ID ClampIntrinsic) {
   return true;
 }
 
+static bool expandSaturateIntrinsic(CallInst *SaturateCall) {
```
wait for @pow2clk  change merges then change this to return `Result` and remove the `replaceAllUsesWith` and `eraseFromParent` calls

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1285,6 +1289,34 @@ static unsigned getBoolCmpOpcode(unsigned PredNum) {
   }
 }
 
+static APFloat getZeroFP(const Type *LLVMFloatTy) {
```
I'm a little suprised this didn't already exist.

Commit:
Add test to verify generation of llvm.spv.saturate.* instructions
Change type check to assert.

farzonl commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -325,6 +325,16 @@ def Abs :  DXILOp<6, unary> {
+  let LLVMIntrinsic = int_dx_saturate;
+  let arguments = [overloadTy];
+  let result = overloadTy;
+  let overloads = [Overloads<DXIL1_0, [halfTy, floatTy, doubleTy]>];
```
I was a little suprised by `double` but yeah dxc generate `double @dx.op.unary.f64(i32 7, double %5)`

farzonl commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -362,6 +364,34 @@ static bool expandClampIntrinsic(CallInst *Orig, Intrinsic::ID ClampIntrinsic) {
+static bool expandSaturateIntrinsic(CallInst *SaturateCall) {
+  FunctionType *FT = SaturateCall->getFunctionType();
+  Type *FTRetTy = FT->getReturnType();
+  assert(FTRetTy == FT->getParamType(0) &&
```
It looks like you are trying to do scalarization via this pass, but we haven't been doing that thus far. I wanted to take care of scalarization in a different pass because doing it here means we have to do it for each intrinsic which I want to avoid. My feeling is scalarization should be uniform across all intrinsics unless there are special cases like `dot2`, `dot3`, `dot4`.

bharadwajy replied:
> It looks like you are trying to do scalarization via this pass, but we haven't been doing that thus far. I wanted to take care of scalarization in a different pass because doing it here means we have to do it for each intrinsic which I want to avoid. My feeling is scalarization should be uniform across all intrinsics unless there are special cases like `dot2`, `dot3`, `dot4`.

The functionality can be moved out of here once such a scalarization pass is implemented.

farzonl replied:
I think we should drop this for now. We likely won't be preserving any intrinsic specific behaviors.

bharadwajy replied:
> I think we should drop this for now. We likely won't be preserving any intrinsic specific behaviors.

Pushed a change dropping scalarization functionality and associated tests.

Commit:
Drop vector operation expansion of saturate as it is expected to
be handled in a scalarization pass planned to be implemented in
the future.

Remove saturate tests with vector operands.

farzonl commented on llvm/test/CodeGen/DirectX/saturate.ll:
```diff
@@ -0,0 +1,69 @@
+; Function Attrs: nocallback nofree nosync nounwind willreturn
+declare double @llvm.dx.saturate.f64(double) #1
+
+attributes #0 = { convergent nounwind "frame-pointer"="all" "no-trapping-math"="true" "stack-protector-buffer-size"="8" }
```
delete lines 52-69?

farzonl commented on llvm/test/CodeGen/DirectX/saturate.ll:
```diff
@@ -0,0 +1,69 @@
+; CHECK-LABEL: test_saturate_double
+define noundef double @test_saturate_double(double noundef %p0) #0 {
+entry:
+  %p0.addr = alloca double, align 8
```
delete lines 40-42? pass `%p0` to `@llvm.dx.saturate.f64` instead of %0. repeat  for the float and half case.

farzonl commented on llvm/test/CodeGen/DirectX/saturate.ll:
```diff
@@ -0,0 +1,69 @@
+; Make sure the intrinsic dx.saturate is to appropriate DXIL op for half/float/double data types.
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
+target triple = "dxilv1.6-unknown-shadermodel6.6-library"
```
your triple via the  opt call is `dxil-pc-shadermodel6.3-library`.  We ussualy either specify the triple here or via the command line to avoid inconsistencies like this. I prefer the command line so recommand deleting this line.

farzonl commented on llvm/test/CodeGen/DirectX/saturate.ll:
```diff
@@ -0,0 +1,69 @@
+; RUN: opt -S -dxil-op-lower -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
+; Make sure the intrinsic dx.saturate is to appropriate DXIL op for half/float/double data types.
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
```
I don't think we need the data layout for this test.

farzonl commented on llvm/test/CodeGen/DirectX/saturate_errors.ll:
```diff
@@ -0,0 +1,14 @@
+; Make sure the intrinsic dx.saturate is to appropriate DXIL op for half/float/double data types.
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
+target triple = "dxilv1.6-unknown-shadermodel6.6-library"
```
same issue here, either delete this line or remove the triple  from the opt flag.

Commit:
Cleanup test sources.
